### PR TITLE
Feature/gh 201 use ssh agent

### LINKED ISSUE #201: Use ssh-agent to not write ssh private keys on disk
**As:** An ops
**I want**: to ensure that connection credentials are not left non-encrypted on disk
**So that:** I can relax and sleep this night

**AC1:** Allow to either specify a key path or a key content into the credential part of a Compute
**AC2:** before running Ansible command store key content into an ssh-agent for a limited amount of time 
**AC3:** check if it also works with paramiko (anyways document what works or not) 
**AC4:** support key content for terraform provisioning
**AC5: Keys content should never be store unencrypted on disk**

### PR DESCRIPTION
**Pull request by stebenoist**
# Pull Request description

**Use ssh-agent to not write ssh private keys on disk** 
As: An ops
I want: to ensure that connection credentials are not left non-encrypted on disk
So that: I can relax and sleep this night

AC1: Allow to either specify a key path or a key content into the credential part of a Compute
AC2: before running Ansible command store key content into an ssh-agent for a limited amount of time
AC3: check if it also works with paramiko (anyways document what works or not)
AC4: support key content for terraform provisioning
AC5: Keys content should never be store unencrypted on disk

## Description of the change

- Add **use_ssh_agent** config option set to true by default.
- Allow using private key file if **use_ssh_agent**=false
- Allow provide private key content for Terraform check connection
- Update doc.

### What I did
**Add ssh_agent API to:**
- create new ssh-agent, add/remove keys, stop...

**Update Ansible to :**
- use ssh-agent if use_ssh_agent=true
- use private key file if use_ssh_agent=false (check if private key is a valid path)

**Update Terraform to:**
- Allow providing private key content for Terraform check connection
- Use **terraform** env var to store private key content

### How to verify it
- Deploy a topology with SSH private key content set in Compute credentials (On demand resource configuration within the A4C Location configuration)
- Config Yorc with use_ssh_agent=false and deploy the same topology with private key file path.
- Test with valid and not valid cerficates. When using  a valid certificate, you can use our home made yorc key pair, or another one created by your own, but in this latter case, you should use a tool that generates the private key in DER format (BER format not suuported)

## Applicable Issues
fixes #201

### CODE REVIEW

**Comment by laurentganne** [comment]
Seeing a failure in unit tests:
```
        --- FAIL: TestRunConsulGooglePackageTests/googleProvider/simpleComputeInstanceWithPersistentDisk (4.43s)
	Error Trace:	
	Error:		Received unexpected error "\"~/.ssh/yorc.pem\" is not a valid path"
	Messages:	Unexpected error attempting to generate compute instance for simpleComputeInstanceWithPersistentDisk
```

Looks like the testdata yaml file describing the compute should be updated to provide a private key content ?

**Comment by stebenoist** [comment]
The file ~/.ssh/yorc.pem is not present in Travis build vm. I fix it with a test key pem file

File: helper/sshutil/sshutil.go [hunk 1]
```diff
@@ -129,6 +143,21 @@ func (sw *SSHSessionWrapper) RunCommand(ctx context.Context, cmd string) error {
 // - either a path to the private key file,
 // - or the content or this private key file
 func ReadPrivateKey(pk string) (ssh.AuthMethod, error) {
+	raw, err := ToPrivateKeyContent(pk)
```
Comment (loicalbertin) [inline helper/sshutil/sshutil.go]: Error is silently ignored....

File: helper/sshutil/sshutil.go [hunk 2]
```diff
@@ -129,6 +143,21 @@ func (sw *SSHSessionWrapper) RunCommand(ctx context.Context, cmd string) error {
...
+
+	signer, err := ssh.ParsePrivateKey(raw)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to parse key file %q", pk)
+	}
+	if err != nil {
```
Comment (loicalbertin) [inline helper/sshutil/sshutil.go]: Looks like the missing error check above

File: helper/stringutil/stringutil.go [hunk 3]
```diff
@@ -76,3 +76,12 @@ func GetFilePath(pathOrContent string) (string, bool, error) {
 
 	return filepath, false, file.Close()
 }
+
+// Truncate truncates a string if it's longer than defined length
+// Add ... after the cut
+func Truncate(str string, l int) string {
+	if len(str) > l {
```
Comment (loicalbertin) [inline helper/stringutil/stringutil.go]: What if l < 0
or l < 3 ?

File: prov/ansible/execution.go [hunk 4]
```diff
@@ -771,9 +773,11 @@ func (e *executionCommon) generateHostConnection(ctx context.Context, buffer *by
 		sshCredentials := e.getSSHCredentials(ctx, host, true)
 		buffer.WriteString(fmt.Sprintf(" ansible_ssh_user=%s ansible_ssh_common_args=\"-o ConnectionAttempts=20\"", sshCredentials.user))
 		// Set with priority private key against password
-		if sshCredentials.privateKey != "" {
-			// TODO if not a path store it somewhere
-			// Note whould be better if we can use it directly https://github.com/ansible/ansible/issues/22382
+		if !e.cfg.UseSSHAgent && sshCredentials.privateKey != "" {
+			// check privateKey's a valid path
+			if is, err := pathutil.IsValidPath(sshCredentials.privateKey); err != nil || !is {
+				return errors.Errorf("%q is not a valid path", stringutil.Truncate(sshCredentials.privateKey, 20))
```
Comment (loicalbertin) [inline prov/ansible/execution.go]: Why truncate where?
We want to will typically want to get the invalid path and make `ls /my/deep/deep/path/to/key.pem ; ls /my/deep/deep/path/to; ls /my/deep/deep/path/; ls /my/deep/deep/` and so on to find the missing part.

File: prov/terraform/google/generator.go [hunk 5]
```diff
@@ -37,11 +38,24 @@ const infrastructureName = "google"
 type googleGenerator struct {
 }
 
-func (g *googleGenerator) GenerateTerraformInfraForNode(ctx context.Context, cfg config.Configuration, deploymentID, nodeName, infrastructurePath string) (bool, map[string]string, []string, error) {
+func getSSHAgent(ctx context.Context, privateKey string) (*sshutil.SSHAgent, error) {
```
Comment (loicalbertin) [inline prov/terraform/google/generator.go]: Why is this not in the commons package?

File: prov/terraform/commons/generator.go [hunk 6]
```diff
@@ -37,7 +37,8 @@ type Generator interface {
 	// GenerateTerraformInfraForNode can also return a map of outputs names indexed by consul keys into which the outputs results should be stored.
 	// And a list of environment variables in form "key=value" to be added to the current process environment when running terraform commands.
 	// This is particularly useful for adding secrets that should not be in tf files.
-	GenerateTerraformInfraForNode(ctx context.Context, cfg config.Configuration, deploymentID, nodeName, infrastructurePath string) (bool, map[string]string, []string, error)
+	// It returns too a callback function allowing execution some cleanup stuff once the infrastructure has been applied
+	GenerateTerraformInfraForNode(ctx context.Context, cfg config.Configuration, deploymentID, nodeName, infrastructurePath string) (bool, map[string]string, []string, func(), error)
```
Comment (loicalbertin) [inline prov/terraform/commons/generator.go]: For clarity I would have make the callback function a named type:

```go
type PostApplyCallback func()

type Generator interface {
 	GenerateTerraformInfraForNode(ctx context.Context, cfg config.Configuration, deploymentID, nodeName, infrastructurePath string) (bool, map[string]string, []string, PostApplyCallback, error)
}
```

File: helper/sshutil/sshutil.go (hunk 2 above)
Comment (stebenoist) [reply helper/sshutil/sshutil.go]: Oups

**Pull request merged**